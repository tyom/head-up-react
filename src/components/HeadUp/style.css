* {
  box-sizing: border-box;
}

body {
  margin: 0;
}

.head-up {
  display: flex;
  font: 14px/1.3 'Helvetica Neue', sans-serif;
  background: #222;
  color: #ddd;
  -webkit-font-smoothing: antialiased;
}

.collection {
  width: 100%;
  height: 100vh;
  overflow: hidden;
}
