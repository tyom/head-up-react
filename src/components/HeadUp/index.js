import React, { Component } from 'react';
import PropTypes from 'prop-types';
import mousetrap from 'mousetrap';
import flatten from 'lodash/flatten';

import Menu from '../Menu';
import Dashboard from '../Dashboard';

import './style.css';

export default class HeadUp extends Component {
  constructor(props) {
    super(props);

    this.state = {
      menuItems: [],
      dashboards: [],
    };
  }

  componentDidMount() {
    const dashboards = this.props.dashboards || flatten([this.props.children]);
    const menuItems = dashboards.map(
      dashboard => dashboard.name || dashboard.props.name
    );
    const getCurrentDashboardIndex = () =>
      menuItems.indexOf(this.props.activeDashboard);

    this.setState({ menuItems, dashboards });

    mousetrap
      .bind('h', this.props.onToggleMenu)
      .bind(['j', 'ctrl+up'], () =>
        this.getPrevDashboardName(getCurrentDashboardIndex())
      )
      .bind(['k', 'ctrl+down'], () =>
        this.getNextDashboardName(getCurrentDashboardIndex())
      );
  }

  componentWillUnmount() {
    mousetrap
      .unbind('h')
      .unbind(['j', 'ctrl+up'])
      .unbind(['k', 'ctrl+down']);
  }

  getPrevDashboardName(currentIndex) {
    const prevIndex =
      currentIndex - 1 < 0 ? this.state.menuItems.length - 1 : currentIndex - 1;
    this.props.onPrevDashboard(this.state.menuItems[prevIndex]);
  }

  getNextDashboardName(currentIndex) {
    const nextIndex =
      currentIndex + 1 > this.state.menuItems.length - 1 ? 0 : currentIndex + 1;
    this.props.onNextDashboard(this.state.menuItems[nextIndex]);
  }

  renderMenu() {
    if (this.state.menuItems.length < 2) {
      return null;
    }
    return (
      <Menu
        items={this.state.menuItems}
        activeDashboard={this.props.activeDashboard}
        isMenuClosed={this.props.isMenuClosed}
        onToggleMenu={this.props.onToggleMenu}
        onSelectMenuItem={this.props.onSelectMenuItem}
      />
    );
  }

  render() {
    return (
      <div styleName="head-up">
        {this.renderMenu()}
        <div styleName="collection">
          {this.props.children.map(({ props }) => (
            <Dashboard
              {...props}
              key={props.name}
              isActive={this.props.activeDashboard === props.name}
            />
          ))}
        </div>
      </div>
    );
  }
}

HeadUp.propTypes = {
  children: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  dashboards: PropTypes.array,
  activeDashboard: PropTypes.string,
  isMenuClosed: PropTypes.bool,
  onToggleMenu: PropTypes.func,
  onSelectMenuItem: PropTypes.func,
  onPrevDashboard: PropTypes.func,
  onNextDashboard: PropTypes.func,
};
