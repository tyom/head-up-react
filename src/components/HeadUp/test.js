import React from 'react';
import mousetrap from 'mousetrap';
import HeadUp from './';

const onToggleMenuSpy = jest.fn();
const onSelectMenuItemSpy = jest.fn();
const onPrevDashboardSpy = jest.fn();
const onNextDashboardSpy = jest.fn();

jest.mock('mousetrap', () => ({
  bind: jest.fn().mockReturnThis(),
  unbind: jest.fn().mockReturnThis(),
}));

function getDashboards(numberOfDashboards = 0, includeInvalid = false) {
  const dashboardsArray = [...Array(numberOfDashboards).keys()].map(i => (
    <div name={`dashboard #${i + 1}`} key={i + 1}>
      <div>cell #{i + 1}.1</div>
      <div>cell #{i + 1}.2</div>
    </div>
  ));

  if (includeInvalid) {
    dashboardsArray.unshift(<div name="dashboard #0" key={0} />);
  }

  return dashboardsArray;
}

const oneDashboard = <HeadUp>{getDashboards(1)}</HeadUp>;
const twoDashboards = <HeadUp>{getDashboards(2)}</HeadUp>;
const twoDashboardsWithInvalid = <HeadUp>{getDashboards(2, true)}</HeadUp>;

describe('HeadUp: structure', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("* doesn't render", () => {
    test('HeadUp when no children are given', () => {
      const component = shallow(<HeadUp />);
      expect(component).toMatchSnapshot();
    });

    test('Menu when only only one dashboard', () => {
      const component = shallow(oneDashboard).find('Menu');
      expect(component).toMatchSnapshot();
    });
  });

  describe('* renders', () => {
    test('dashboards with children', () => {
      const component = shallow(twoDashboards).find('Dashboard');
      expect(component).toMatchSnapshot();
    });

    test('dashboards ignoring invalid', () => {
      const dashboards = shallow(twoDashboardsWithInvalid).find('Dashboard');
      expect(dashboards).toHaveLength(2);
    });

    test('renders Menu with dashboard layout', () => {
      const component = shallow(twoDashboards).find('Menu');
      expect(component).toMatchSnapshot();
    });
  });
});

describe('HeadUp: behaviour', () => {
  describe('* navigation', () => {
    let component;

    const threeDashboardsSecondActive = (
      <HeadUp
        activeDashboard="dashboard #2"
        onPrevDashboard={onPrevDashboardSpy}
        onNextDashboard={onNextDashboardSpy}
        onSelectMenuItem={onSelectMenuItemSpy}
        onToggleMenu={onToggleMenuSpy}
      >
        {getDashboards(3)}
      </HeadUp>
    );

    beforeEach(() => {
      jest.clearAllMocks();
      component = shallow(threeDashboardsSecondActive);
    });

    test('getting current dashboard index', () => {
      const index = component.instance().getCurrentIndex();

      expect(index).toEqual(1);
    });

    test('getting previous dashboard name', () => {
      const name = component.instance().getPrevDashboardName();

      expect(name).toEqual('dashboard #1');
    });

    test('getting next dashboard name', () => {
      const name = component.instance().getNextDashboardName();

      expect(name).toEqual('dashboard #3');
    });

    test('calling #onPrevDashboard prop with previous dashboard name', () => {
      component.instance().selectPrevDashboard();

      expect(onPrevDashboardSpy).toBeCalledWith('dashboard #1');
    });

    test('calling #onPrevDashboard prop with the last dashboard name after the first (cycle)', () => {
      component.setProps({ activeDashboard: 'dashboard #1' });
      component.instance().selectPrevDashboard();

      expect(onPrevDashboardSpy).toBeCalledWith('dashboard #3');
    });

    test('calling #onNextDashboard prop', () => {
      component.instance().selectNextDashboard();

      expect(onNextDashboardSpy).toBeCalled();
    });

    test('calling #onNextDashboard prop with the first dashboard name after the last (cycle)', () => {
      component.setProps({ activeDashboard: 'dashboard #3' });
      component.instance().selectNextDashboard();

      expect(onNextDashboardSpy).toBeCalledWith('dashboard #1');
    });


    test('binding keyboard shortcuts', () => {
      expect(mousetrap.bind.mock.calls).toEqual([
        ['h', expect.any(Function)],
        [['j', 'ctrl+up'], expect.any(Function)],
        [['k', 'ctrl+down'], expect.any(Function)],
      ]);
    });

    test('unbinding keyboard shortcuts', () => {
      component.instance().componentWillUnmount();

      expect(mousetrap.unbind.mock.calls).toEqual([
        ['h'],
        [['j', 'ctrl+up']],
        [['k', 'ctrl+down']],
      ]);
    });
  });
});
