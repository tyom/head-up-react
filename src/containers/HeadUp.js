import { connect } from 'react-redux';

import { activateDashboard, toggleMenu } from '../actions';
import HeadUp from '../components/HeadUp';

const mapStateToProps = (state, ownProps) => ({
  cells: ownProps.children,
  activeDashboard: state.headup.activeDashboard,
  isMenuClosed: state.headup.isMenuClosed,
});

const mapDispatchToProps = dispatch => ({
  onToggleMenu: () => dispatch(toggleMenu()),
  onSelectMenuItem: name => dispatch(activateDashboard(name)),
  onPrevDashboard: name => dispatch(activateDashboard(name)),
  onNextDashboard: name => dispatch(activateDashboard(name)),
});

export default connect(mapStateToProps, mapDispatchToProps)(HeadUp);
