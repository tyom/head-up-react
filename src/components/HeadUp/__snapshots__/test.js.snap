// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadUp: structure * doesn't render HeadUp when no children are given 1`] = `""`;

exports[`HeadUp: structure * doesn't render Menu when only only one dashboard 1`] = `null`;

exports[`HeadUp: structure * renders dashboards with children 1`] = `
Array [
  <Dashboard
    isActive={false}
    key="dashboard #1"
    name="dashboard #1"
  >
    <div>
      cell #
      1
      .1
    </div>
    <div>
      cell #
      1
      .2
    </div>
  </Dashboard>,
  <Dashboard
    isActive={false}
    key="dashboard #2"
    name="dashboard #2"
  >
    <div>
      cell #
      2
      .1
    </div>
    <div>
      cell #
      2
      .2
    </div>
  </Dashboard>,
]
`;

exports[`HeadUp: structure * renders renders Menu with dashboard layout 1`] = `
<Menu
  items={
    Array [
      Object {
        "cells": Array [
          undefined,
          undefined,
        ],
        "name": "dashboard #1",
      },
      Object {
        "cells": Array [
          undefined,
          undefined,
        ],
        "name": "dashboard #2",
      },
    ]
  }
/>
`;
